lib/rust/parser/cdylib/src/lib.rs

```diff
@@ -0,0 +1,224 @@
+///     const uintptr_t output_len = parse_result_get_output_data_len(result);
+///     // Do something with the results.
+///     handle_parse_result(sanitized_input || input, output, output_len);
+///     // Free resources.
+///     parse_result_free(result);
+///     parser_free(parser);
+/// }
+/// ```
+#[derive(Debug, Default)]
+#[allow(missing_copy_implementations)] // Expected to contain non-Copy data in the future.
+pub struct Parser {
+    // The implementation doesn't currently support reusing parser resources, but when we add that
+    // optimization, the API supports it.
+}
+
+
+// === Implementation ===
+
+impl Parser {
+    fn run(&mut self, input: &[u8]) -> ParseResult {
+        let input = String::from_utf8_lossy(input);
+        let tree = enso_parser::Parser::new().run(&input);
+        let tree = match enso_parser::serialization::serialize_tree(&tree) {
+            Ok(tree) => tree,
+            // `Tree` does not contain any types with fallible `serialize` implementations, so this
+            // cannot fail.
+            Err(_) => default(),
+        };
+        let input = match input {
+            Cow::Owned(input) => Some(input),
+            _ => None,
+        };
+        ParseResult { input, tree }
+    }
+}
+
+
+// === C Interface ===
+
+/// Create a new parser. This operation is infallible; the result will never be NULL.
+#[no_mangle]
+pub extern "C" fn parser_new() -> *mut Parser {
+    Box::into_raw(default())
+}
+
+/// Parse the given source code, provided in the input parameters. The result will be written to the
+/// object provided in the output parameter; any previous state of the output object will be
+/// overwritten.
+///
+/// # Safety
+///
+/// The `parser` argument must be a value returned by `parser_new` that has never been passed to
+/// `parser_free`.
+/// The `input` argument must be a non-NULL pointer to a sequence of at least `input_len` readable
+/// bytes.
+/// The `output` argument must be a value returned by `parse_result_alloc` that has never been
+/// passed to `parse_result_free`.
+#[no_mangle]
+pub unsafe extern "C" fn parser_run(
+    parser: *mut Parser,
+    input: *const u8,
+    input_len: usize,
+    output: *mut ParseResult,
+) {
+    let parser = &mut *parser;
+    let input = slice::from_raw_parts(input, input_len);
+    let output = &mut *output;
+    *output = parser.run(input);
+}
+
+/// Release the parser's resources.
+///
+/// # Safety
+///
+/// The `parser` argument must be a value returned by `parser_new` that has never been passed to
+/// `parser_free`.
+#[no_mangle]
+pub unsafe extern "C" fn parser_free(parser: *mut Parser) {
+    let parser = Box::from_raw(parser);
+    drop(parser);
+}
+
+
+
+// ===================
+// === ParseResult ===
+// ===================
+
+/// The result of parsing. Contains the source code, if the parser modified it, and the serialized
+/// AST.
+#[derive(Debug, Default)]
+pub struct ParseResult {
+    input: Option<String>,
```

Comment:
How does this `input` differ from the `input` in `parser_run` function?

Reply:
The fields are documented on their accessors (`parse_result_get_input_data`). The parser can only process UTF-8. If the input is not UTF-8, we attempt to recover by replacing all sequences of bytes that can't be read as UTF-8 with a placeholder codepoint. In that case, all source code references will be relative to this modified input.

Reply:
On the Java side I will load a `String` and convert it to `ByteBuffer` in UTF-8 format - e.g. the input will always be UTF-8. Btw. another problem is that I will need indexes in characters, not bytes...